README.md
# GitVersionDemo

SOME NEW INFORMATION


CHANGELOG.md
# 0.0.2

UPDATE

# 0.0.21

make UPDATE lower case (T222 is fixed)


scripts/build.py
import json
import pathlib
import platform
import re
import subprocess
import sys
from git.repo import Repo

from build_logger import logger


# {major}.{minor}.{patch}-{tag}+{buildmetadata}

def _exec(command: str) -> str:
    code, output = subprocess.getstatusoutput(command)
    if code != 0:
        raise OSError(f'Fail to execute {command} with code {code}, {output}')
    return output


class Builder(object):

    def build(self) -> None:
        """入口"""

        root = pathlib.Path(__file__).parent.parent
        version: str = json.load(open(root.joinpath('package.json')))['version']
        if not re.fullmatch(r'[0-9]+\.[0-9]+\.[0-9]+', version):
            raise ValueError('package.json 中的 version 字段必须是 {major}.{minor}.{patch}, 并且每个关键词都是自然数')

        repo = Repo(root.joinpath('.git'))
        major, minor, patch = version.split(r'.')
        logger.info('Start building Modeling & Simulation Studio...')
        logger.debug('================= System Info =================')
        logger.debug(f'Platform Info  | {platform.platform()}')
        logger.debug(f'System Release | {platform.version()}')
        logger.debug(f'Node Name      | {platform.node()}')

        python_version = sys.version.replace('\n', '')
        logger.debug('================= Environment =================')
        logger.debug(f'* NodeJS {_exec("node --version")}')
        logger.debug(f'* g++ {_exec("g++ --version")}')
        logger.debug(f'* CMake {_exec("cmake --version")}')
        logger.debug(f'* Python {python_version}')

        current_commit = repo.head.commit
        sha = current_commit.hexsha

        branch_name = repo.active_branch.name
        logger.debug('===================== Git ======================')
        logger.debug(f'* branch={branch_name}')
        logger.debug(f'* sha={sha}')

        pre_release: str
        build_meta: str
        if branch_name == 'develop':
            pre_release = 'alpha'
            build_meta = repo.git.rev_list('--count', f'{repo.heads["main"].commit.hexsha}..{current_commit}')
        elif branch_name.startswith('release') or branch_name.startswith('hotfix'):
            pre_release = 'beta'
            build_meta = repo.git.rev_list('--count', f'{repo.heads["develop"].commit.hexsha}..{current_commit}')
        elif re.fullmatch(r'feature/T[0-9]+', branch_name) or re.fullmatch(r'fix/T[0-9]+', branch_name):
            manifest = branch_name.split('/')
            build_meta = repo.git.rev_list('--count', f'{repo.heads["develop"].commit.hexsha}..{current_commit}')
            pre_release = manifest[1]
        else:
            raise NotImplementedError(f'Not Support build branch {branch_name}')

        semantic_version = f'{major}.{minor}.{patch}-{pre_release}.{build_meta}'
        print('semantic_version', semantic_version)

builder = Builder()

builder.build()


scripts/git_utils.py
import logging
import pathlib
import typing
import re

def get_git_version(repo: pathlib.Path) -> typing.Optional[str]:
    git = pathlib.Path(repo, '.git')
    head_path = git.joinpath('HEAD')
    
    if not head_path.exists():
        logging.warning(f'{head_path.as_posix()} not exists, {repo.as_posix()} is not a valid git repo')
        return

    head = open(head_path).read()

    if re.fullmatch(r'^[0-9a-f]{40}$', head):
        return head

    ref_match = re.findall(r'^ref: (.*)$', head)

    if not ref_match:
        return None

    ref = ref_match[0]
    ref_path = git.joinpath(ref)

    try:
        return open(ref_path, encoding='utf-8').read().strip()
    except:
        pass
    # packed_refs_path = git.joinpath('packed-refs')

    # refs_raw = ''
    # try:
    #     refs_raw = open(packed_refs_path, encoding='utf-8').read().strip()
    # except:
    #     return None

    # refs_regex = 


    return None

scripts/build_logger.py
import logging
import sys

import typing


class _BuildLoggerFormatter(logging.Formatter):
    grey = '\x1b[38;20m'
    light_green = '\x1b[92;20m'
    yellow = '\x1b[33;20m'
    red = '\x1b[31;20m'
    bold_red = '\x1b[31;1m'
    reset = '\x1b[0m'
    format = '[%(levelname)s][%(asctime)s] %(message)s'

    FORMATS = {
        logging.DEBUG: grey + format + reset,
        logging.INFO: light_green + format + reset,
        logging.WARNING: yellow + format + reset,
        logging.ERROR: red + format + reset,
        logging.CRITICAL: bold_red + format + reset
    }

    def format(self, record: logging.LogRecord) -> str:
        log_fmt = self.FORMATS.get(record.levelno)
        formatter = logging.Formatter(log_fmt)
        return formatter.format(record)


logger: typing.Final[logging.Logger] = logging.getLogger("Builder")
logger.setLevel(logging.DEBUG)

# create console handler with a higher log level
_console_handler = logging.StreamHandler(stream=sys.stdout)
_console_handler.setLevel(logging.DEBUG)

_console_handler.setFormatter(_BuildLoggerFormatter())

logger.addHandler(_console_handler)


scripts/git_repository.py
import pathlib
import typing
import re


class GitRepository(object):
    """Git 仓库管理"""
    def __init__(self, repo_path: pathlib.Path) -> None:
        self.__repo_path = repo_path
        self.__git = pathlib.Path(self.__repo_path, '.git')

    @property
    def repo_path(self) -> pathlib.Path:
        return self.__repo_path

    def get_current_branch(self) -> str:
        """获取当前 HEAD 的分支名称

        Raises:
            ValueError: 如果当前 HEAD 不是分支而是 commit, raise

        Returns:
            str: 当前 HEAD 的分支名称
        """
        head = self.__get_current_head()
        if re.fullmatch(r'^[0-9a-f]{40}$', head):
            raise ValueError(f'Current HEAD {head} is only commit, not on any branch')

        ref_match = re.findall(r'^ref: (.*)$', head)
        return ref_match[0]

    def get_current_commit(self) -> typing.Optional[str]:
        """获取当前 HEAD 的 commit 号

        Returns:
            typing.Optional[str]: 当前 HEAD 的 commit 号, 如果找不到返回 None
        """
        head = self.__get_current_head()

        if re.fullmatch(r'^[0-9a-f]{40}$', head):
            return head

        ref_match = re.findall(r'^ref: (.*)$', head)

        if not ref_match:
            return None

        return self.__get_commit_by_ref(ref_match[0])

    def get_commit_by_branch(self, branch_name: str) -> str:
        """获取分支上的最新 commit 号

        Args:
            branch_name (str): 需要检查的分支名称

        Returns:
            str: 该分支上的最新 commit 号
        """
        return self.__get_commit_by_ref(f'refs/heads/{branch_name}')

    def __get_commit_by_ref(self, ref: str) -> str:
        if not ref.startswith('refs/heads'):
            raise ValueError(f'ref must starts with "refs/heads", given {ref}')
        return open(self.__git.joinpath(ref), encoding='utf-8').read().strip()
       
    def __get_current_head(self) -> str:
        head_path = self.__git.joinpath('HEAD')
        
        if not head_path.exists():
            raise FileNotFoundError(f'{head_path.as_posix()} not exists, {self.__repo_path.as_posix()} is not a valid git repo_path')

        return open(head_path).read()


scripts/release.py
# -*- coding: utf-8 -*-
"""
:文件名: scripts/release.py

:创建时间: 03/04/2022 10:50 AM

:作者: 许翀轶

:模块描述: release 脚本帮助

*Copyright (c) 2021 上海博佳医药科技有限公司*
"""
import argparse
import logging
import pathlib
import json
import pathlib
import re
import sys
import typing
import time
import git
import git.repo as git_repo


class _ConsoleLoggerFormatter(logging.Formatter):
    grey = '\x1b[38;20m'
    light_green = '\x1b[92;20m'
    yellow = '\x1b[33;20m'
    red = '\x1b[31;20m'
    bold_red = '\x1b[31;1m'
    reset = '\x1b[0m'
    record_fmt = '[%(levelname)s][%(asctime)s] %(message)s'

    FORMATS = {
        logging.DEBUG: record_fmt,
        logging.INFO: light_green + record_fmt + reset,
        logging.WARNING: yellow + record_fmt + reset,
        logging.ERROR: red + record_fmt + reset,
        logging.CRITICAL: bold_red + record_fmt + reset
    }

    def format(self, record: logging.LogRecord) -> str:
        log_fmt = self.FORMATS.get(record.levelno)
        formatter = logging.Formatter(log_fmt)
        return formatter.format(record)


def _round(value: float, decimal: int = 3) -> float:
    return round(value, decimal)


def release(release_type: typing.Literal['patch', 'minor'], ignore_uncommitted_changes: bool = False) -> None:
    """release 支持"""
    logger: typing.Final[logging.Logger] = logging.Logger("MaSReleaser", level=logging.DEBUG)
    # create console handler
    _console_handler = logging.StreamHandler(stream=sys.stdout)
    _console_handler.setFormatter(_ConsoleLoggerFormatter())
    logger.addHandler(_console_handler)

    
    try:
        root = pathlib.Path(__file__).parent.parent
        sys.path.append(root.as_posix())

        package_json_fp = root.joinpath('package.json')
        package_json: typing.Final[typing.Dict[str, typing.Any]] = json.load(open(package_json_fp, encoding='utf-8'))
        app_version = package_json['version']
        if not re.fullmatch(r'[0-9]+\.[0-9]+\.[0-9]+', app_version):
            raise ValueError('package.json 中的 version 字段必须是 {major}.{minor}.{patch}, 并且每个关键词都是自然数')

        logger.info('================== Arguments ==================')
        logger.info(f'* type={release_type}')
        logger.info(f'* ignore_uncommitted_changes={ignore_uncommitted_changes}')

        repo = git_repo.Repo(root.joinpath('.git'))
        if repo.is_dirty(untracked_files=True):
            if not ignore_uncommitted_changes:
                raise ValueError('Uncommitted changes found, make sure you have committed all changes')
            logger.warning(f'Uncommitted changes found but is surpressed with ignore_uncommitted_changes=True')

        start_time = time.time()
        logger.info(f'Pulling from origin {repo.remotes.origin.url}...')
        repo.remotes.origin.pull()
                
        major, minor, patch = app_version.split(r'.')
        current_commit = repo.head.commit
        branch_name = repo.active_branch.name
        sha = current_commit.hexsha
        author_info = f'{current_commit.author.name}<{current_commit.author.email}>'
        datetime_info = repo.head.commit.committed_datetime.strftime('%Y-%m-%d %H:%M:%S%z')
        logger.debug('=================== Git =====================')
        logger.debug(f'* branch={branch_name}')
        logger.debug(f'* sha={sha}')
        logger.debug(f'* author_info={author_info}')
        logger.debug(f'* datetime_info={datetime_info}')

        def _branch_exists_on_remote(branch_name: str) -> bool:
            for ref in typing.cast(typing.Iterable[git.RemoteReference], repo.remote().refs):
                if branch_name in ref.name:
                    return True

            return False

        def _change_version(version: str) -> None:
            package_json['version'] = version
            logger.info(f'Versioning {version}...')
            open(package_json_fp, mode='w', encoding='utf-8').write(json.dumps(package_json, ensure_ascii=True, indent=4))
            

        if release_type == 'patch':
            semantic_version = f'{major}.{minor}.{int(patch) + 1}'
            hotfix_branch_name = f'hotfix/v{semantic_version}'
            logger.info(f'Starting new hotfix branch {hotfix_branch_name}...')
            if branch_name != 'master':
                raise RuntimeError('You are **MUST** working on "master" branch when you are trying to start a hotfix branch')
            
            if _branch_exists_on_remote(hotfix_branch_name):
                raise RuntimeError(f'{hotfix_branch_name} is already exists, stop release')
            hotfix_branch_head = repo.create_head(hotfix_branch_name)
            hotfix_branch_head.checkout()

            _change_version(semantic_version)
            
            repo.git.add(A=True)
            repo.git.commit(m=f'hotfix: Branching hotfix(v{semantic_version}) from master(v{major}.{minor}.{patch})\n\n@bypass-review')
            logger.info(f'Pushing hotfix branch {hotfix_branch_name}...')
            repo.git.push('--set-upstream', 'origin', hotfix_branch_head)
        elif release_type == 'minor':
            semantic_version = f'{major}.{minor}.{patch}'

            release_branch_name = f'release/v{semantic_version}'
            logger.info(f'Starting new release branch {release_branch_name}...')
            if branch_name != 'develop':
                raise RuntimeError('You are **MUST** working on "develop" branch when you are trying to start a release branch')

            if _branch_exists_on_remote(release_branch_name):
                raise RuntimeError(f'{release_branch_name} is already exists, stop release')

            release_branch_head = repo.create_head(release_branch_name)
            release_branch_head.checkout()
            logger.info(f'Pushing release branch {release_branch_head}...')
            repo.git.push('--set-upstream', 'origin', release_branch_head)

            logger.info(f'Switch back to develop')
            repo.git.checkout('develop')

            _change_version(f'{major}.{int(minor) + 1}.{patch}')
            repo.git.add(A=True)
            repo.git.commit(m=f'develop: Development(v{semantic_version})\n\n@bypass-review')
            logger.info(f'Pushing develop...')
            repo.git.push('origin', 'develop')
        else:
            raise NotImplementedError(f'Release type {release_type} is not supported')

        end_time = time.time()
        logger.info(f'Release finished in {_round(end_time - start_time)}s')
        logger.info('ALL DONE!')
        exit(0)
    except Exception as e:
        logger.error('=================== ERROR ====================')
        logger.error('RELEASE FAILED', exc_info=e)
        exit(1)


if __name__ == '__main__':
    argparser = argparse.ArgumentParser('demo arg parser')
    argparser.add_argument('release_type', type=str, choices=['minor', 'patch'], help='Type to release')
    args = argparser.parse_args()
    release(args.release_type)
